dspy/teleprompt/random_search.py

```diff
@@ -49,27 +55,42 @@ def __init__(
         self.max_num_samples = max_bootstrapped_demos
         self.max_errors = max_errors
         self.num_candidate_sets = num_candidate_programs
-        # self.max_num_traces = 1 + int(max_bootstrapped_demos / 2.0 * self.num_candidate_sets)
+        # if needed: self.max_num_traces=1+int(max_bootstrapped_demos/2.0*self.num_candidate_sets)
 
         # Semi-hacky way to get the parent class's _bootstrap function to stop early.
-        # self.max_bootstrapped_demos = self.max_num_traces
+        # just in case this is needed: self.max_bootstrapped_demos = self.max_num_traces
         self.max_labeled_demos = max_labeled_demos
 
         dspy.logger.info(
-            "Going to sample between", self.min_num_samples, "and", self.max_num_samples, "traces per predictor.",
+            "Going to sample between",
+            self.min_num_samples,
+            "and",
+            self.max_num_samples,
+            "traces per predictor.",
         )
         dspy.logger.info("Will attempt to train", self.num_candidate_sets, "candidate sets.")
 
-    def compile(self, student, *, teacher=None, trainset, valset=None, restrict=None, labeled_sample=True):
+    def compile(self, student, *, teacher=None, trainset, valset=None, restrict=None, labeled_sample=True) -> Any:
         self.trainset = trainset
         self.valset = valset or trainset  # TODO: FIXME: Note this choice.
 
         scores = []
         all_subscores = []
         score_data = []
 
+        if teacher:
+            has_programs = hasattr(teacher, "candidate_programs") or hasattr(teacher, "programs")
+            if hasattr(teacher, "candidate_programs"):
+                teacher.candidate_programs = teacher.candidate_programs[:1]
+            elif hasattr(teacher, "programs"):
+                teacher = teacher.programs[0]
+            if has_programs:
+                # if we are working with a teacher that already has candidate programs
+                # lets skip the zero shot and LabeledFewShot cases since they don't use teachers
+                restrict = [-1]
+
         for seed in range(-3, self.num_candidate_sets):
-            if (restrict is not None) and (seed not in restrict):
+            if seed < 0 and (restrict is not None) and (seed not in restrict):
```

Comment:
why are we skipping over the zero-shot, labeled_fewshot and unshuffled_fewshot cases for all scenarios? I think even in the conditions above when a teacher is set, it is fine to run through these preliminary candidates and have it weighted in the scores.